tests/lang/s02_core/s99_coroutines/t10_self_parameter.nim

```diff
@@ -0,0 +1,17 @@
+discard """
+  output: "Coroutine[void]\ncsRunning"
+"""
+
+## To access the active *instance* within the body of a coroutine, the hidden
+## ``self`` parameter is made available.
+
+# XXX: not a good solution, either the ``self`` parameter should be explicit
+#      (somehow), or there should be a magic procedure
```

Comment:
i can suggest to declare self explicitly
```
proc coro(self: Coroutine) {.coroutine.} =
  echo typeof(self)
  echo self.status
```

Reply:
I don't know if it is, `proc c(a: int) {.coroutine.}` describes at least two things, one is the coroutine constructor, and the other is all the fields (parameter list) that are part of the backing closure environment. The body of `c` assumes both `self` and an unpacked `a`.

We could have `proc c(self: Coroutine, a: int) {.coroutine.}`, but besides being redundant, it only provides an optional rename of the `self` parameter and specification of an alternative Coroutine base type. With that said:

1. `self`'s type isn't `Coroutine`, that's actually the base type
2. renaming `self` is likely _not_ helpful for most readers (misfeature)

If you really want to be explicit, then it'd be more like this:

```nim
proc coro(#[constructor params go here]#) {.coroutine.} =
  proc (self: Coroutine) =
    echo typeof(self)
    echo self.status
  # `coroutine` pragma makes the return type a `void proc closure/Coroutine[void]`
```

Reply:
I had considered this, but I'm not sure if it's better. The first parameter having special meaning seems inconsistent with the rest of the language, where parameter position doesn't have special meaning.

I think it's also somewhat confusing for the programmer:
```nim
proc coro(self: Coroutine) {.coroutine.} =
  discard

# why is there no argument passed to the coroutine constructor?
discard launch(coro())
```

Reply:
> if we're not going to have colorless functions

I'm not sure I understand this remark. The feature as currently specified doesn't introduce "color", in that the coroutines can be launched and resumed from everywhere, without tainting the caller.